File: ims-web/src/sales-invoice/salesRateResolver.ts
import { fetchLatestPurchaseInvoiceSalesRate } from '../api/purchaseInvoices';
import { getSalesPurchaseSettings } from '../api/salesPurchaseSettings';
import { lookupProduct } from '../api/products';
import { findLocalProduct, type SalesProductInfo } from '../components/transaction/salesProductPicker';

export interface SalesRateResolution {
  rate: number;
  warningMessage: string | null;
  usedFallback: boolean;
}

function pickWithFallback(
  primary: number,
  fallback: number,
  productCode: string,
  primaryLabel: string,
  fallbackLabel: string,
): SalesRateResolution {
  if (primary > 0) {
    return { rate: primary, warningMessage: null, usedFallback: false };
  }

  if (fallback > 0) {
    return {
      rate: fallback,
      warningMessage: `No sales rate in ${primaryLabel} for ${productCode}. Using ${fallbackLabel} rate instead.`,
      usedFallback: true,
    };
  }

  return {
    rate: 0,
    warningMessage: `No sales rate found for ${productCode} in product master or purchase invoices. Enter the rate manually.`,
    usedFallback: false,
  };
}

async function resolveProductMasterRate(
  productCode: string,
  productHint?: SalesProductInfo | null,
): Promise<number> {
  if (productHint && productHint.rate > 0) return productHint.rate;

  try {
    const product = await lookupProduct(productCode);
    if (product && product.rate > 0) return product.rate;
  } catch {
    // fall through
  }

  const local = findLocalProduct(productCode);
  if (local && local.rate > 0) return local.rate;

  return 0;
}

async function resolvePurchaseInvoiceRate(productCode: string): Promise<number> {
  try {
    const dto = await fetchLatestPurchaseInvoiceSalesRate(productCode);
    if (!dto?.salesRate) return 0;

    const rate = Number(String(dto.salesRate).replace(/,/g, ''));
    return Number.isFinite(rate) && rate > 0 ? rate : 0;
  } catch {
    return 0;
  }
}

/** WPF: SalesRateResolver.ResolveForSalesInvoiceAsync */
export async function resolveForSalesInvoice(
  productCode: string,
  productHint?: SalesProductInfo | null,
): Promise<SalesRateResolution> {
  const settings = await getSalesPurchaseSettings();
  const masterRate = await resolveProductMasterRate(productCode, productHint);
  const purchaseRate = await resolvePurchaseInvoiceRate(productCode);

  if (settings.salesRateSource === 'purchase_invoice') {
    return pickWithFallback(
      purchaseRate,
      masterRate,
      productCode,
      'purchase invoice',
      'product master',
    );
  }

  return pickWithFallback(
    masterRate,
    purchaseRate,
    productCode,
    'product master',
    'purchase invoice',
  );
}

File: IMS/Services/SalesRateResolver.cs
using IMS.Models;
using IMS.Services.Api;
using IMS.Services.Api.Dtos;

namespace IMS.Services;

public sealed record SalesRateResolution(
    decimal Rate,
    string? WarningMessage,
    bool UsedFallback);

public static class SalesRateResolver
{
    public static async Task<SalesRateResolution> ResolveForSalesInvoiceAsync(
        string productCode,
        SalesProductInfo? productHint = null)
    {
        var source = SalesPurchaseSettingsService.Instance.Current.SalesRateSource;
        var masterRate = await ResolveProductMasterRateAsync(productCode, productHint);
        var purchaseRate = await ResolvePurchaseInvoiceRateAsync(productCode);

        if (source == SalesRateSource.ProductMaster)
            return PickWithFallback(masterRate, purchaseRate, productCode, primaryLabel: "product master", fallbackLabel: "purchase invoice");

        return PickWithFallback(purchaseRate, masterRate, productCode, primaryLabel: "purchase invoice", fallbackLabel: "product master");
    }

    private static SalesRateResolution PickWithFallback(
        decimal primary,
        decimal fallback,
        string productCode,
        string primaryLabel,
        string fallbackLabel)
    {
        if (primary > 0)
            return new SalesRateResolution(primary, null, false);

        if (fallback > 0)
            return new SalesRateResolution(
                fallback,
                $"No sales rate in {primaryLabel} for {productCode}. Using {fallbackLabel} rate instead.",
                true);

        return new SalesRateResolution(
            0,
            $"No sales rate found for {productCode} in product master or purchase invoices. Enter the rate manually.",
            false);
    }

    private static async Task<decimal> ResolveProductMasterRateAsync(string productCode, SalesProductInfo? productHint)
    {
        if (productHint is not null)
        {
            if (productHint.SaleRate > 0)
                return productHint.SaleRate;
            if (productHint.Rate > 0)
                return productHint.Rate;
        }

        if (ImsApiClient.IsAvailable)
        {
            try
            {
                var dto = await ImsApiClient.GetProductByCodeAsync(productCode);
                if (dto is not null && dto.SalePrice > 0)
                    return dto.SalePrice;
            }
            catch
            {
                // fall through
            }
        }

        var local = SalesProductLookup.FindLocal(productCode);
        if (local is not null)
            return local.SaleRate > 0 ? local.SaleRate : local.Rate;

        return 0;
    }

    private static async Task<decimal> ResolvePurchaseInvoiceRateAsync(string productCode)
    {
        if (!ImsApiClient.IsAvailable)
            return 0;

        try
        {
            var dto = await ImsApiClient.GetLatestPurchaseInvoiceSalesRateAsync(productCode);
            if (dto?.SalesRate is null)
                return 0;

            return decimal.TryParse(dto.SalesRate.Replace(",", string.Empty), out var rate) && rate > 0
                ? rate
                : 0;
        }
        catch
        {
            return 0;
        }
    }
}
